Source/EasyNetQ/Consumer/InternalConsumer.cs

```diff
@@ -332,21 +332,19 @@ private void SingleBasicConsumerCancelled(BasicConsumer consumer)
         public void Dispose()
         {
             if (disposed) return;
+
             disposed = true;
 
             var model = Model;
             if (model != null)
             {
                 // Queued because we may be on the RabbitMQ.Client dispatch thread.
-                var disposedEvent = new AutoResetEvent(false);
```

Comment:
Why this code was removed?

Reply:
Please look at this issue https://github.com/EasyNetQ/EasyNetQ/issues/766

Reply:
I think now it is some sort of hand made `IAsyncDisposable` :)

Reply:
Forgive me my paranoia - a potential deadlock is also possible if, for some reason, the dispatcher thread cannot execute the queued delegate, for example unhandled exception from `logger.ErrorException(string.Empty, exception);`

Reply:
Or even simpler - exception in lines
```
 foreach (var c in basicConsumers)
                            c.Dispose();
                        model.Dispose();
```

so `finally` must be here

Reply:
Also maybe worth place `logger.ErrorException(string.Empty, exception);` inside one more `try` with empty `catch` to prevent dispather thread hung.

Reply:
@sungam3r Can you make gist of how you think it should look ? :)

Reply:
```
catch (Exception exception)
{
  try
  {
    logger.ErrorException(string.Empty, exception);
  }
  catch
  {
    // ignore all exceptions if any 
  }
}
```

Reply:
```c#
 consumerDispatcher.QueueAction(() =>
                    {
                        try
                        {
                          foreach (var c in basicConsumers)
                              c.Dispose();
                          model.Dispose();
                       }
                       finally
                       {
                          disposedEvent.Set();
                       }
                    });
```